src/libcrun/linux.c

```diff
@@ -3341,20 +3332,34 @@ init_container (libcrun_container_t *container, int sync_socket_container, struc
         return ret;
     }
 
+  mount_fds = make_libcrun_fd_map (def->mounts_len);
+
   /* If the container needs to join an existing PID namespace, take a reference to it
      before creating a new user namespace, as we could lose the access to the existing
-     namespace.  */
+     namespace.
+
+     This cannot be done in the parent process (by `prepare_and_send_mounts`), since it is
+     necessary to first join the target namespaces and then create the mount.
+*/
   if ((init_status->all_namespaces & CLONE_NEWUSER) && (init_status->join_pidns || init_status->join_ipcns))
     {
       for (i = 0; i < def->mounts_len; i++)
         {
+          int fd = -1;
           /* If for any reason the mount cannot be opened, ignore errors and continue.
              An error will be generated later if it is not possible to join the namespace.
           */
           if (init_status->join_pidns && strcmp (def->mounts[i]->type, "proc") == 0)
-            procfsfd = fsopen_mount (def->mounts[i]);
+            fd = fsopen_mount (def->mounts[i]);
           if (init_status->join_ipcns && strcmp (def->mounts[i]->type, "mqueue") == 0)
-            mqueuefsfd = fsopen_mount (def->mounts[i]);
+            fd = fsopen_mount (def->mounts[i]);
+
+          if (fd >= 0)
+            {
+              if (mount_fds->fds[i] >= 0)
```

Comment:
Could this be ever true since we just created the map.

Reply:
No, I have added it just to be safe in case the code is moved around. I will drop the check

Reply:
sorry my mistake.

`get_fd_map` might return an already existing map, so we are not sure at this point whether it already contains an open file descriptor